gum/backend-darwin/gumprocess-darwin.c

```diff
@@ -347,6 +361,94 @@ gum_process_enumerate_ranges (GumPageProtection prot,
 }
 
 void
+gum_process_enumerate_malloc_ranges (GumFoundMallocRangeFunc func,
+                                     gpointer user_data)
+{
+  task_t task;
+  kern_return_t ret;
+  unsigned i;
+  vm_address_t * malloc_zone_addresses;
+  unsigned malloc_zone_count;
+
+  task = mach_task_self ();
+
+  ret = malloc_get_all_zones (task,
+      gum_read_malloc_memory, &malloc_zone_addresses,
+      &malloc_zone_count);
+  if (ret != KERN_SUCCESS)
+    return;
+
+  for (i = 0; i != malloc_zone_count; i++)
+  {
+    vm_address_t zone_address;
+    malloc_zone_t * zone;
+
+    zone_address = malloc_zone_addresses[i];
+    zone = (malloc_zone_t *) zone_address;
+
+    if (zone && zone->introspect && zone->introspect->enumerator) {
+      GumEnumerateMallocRangesContext info = { func, user_data, TRUE };
+
+      zone->introspect->enumerator (task, (void *) &info,
+          MALLOC_PTR_IN_USE_RANGE_TYPE, zone_address,
+          gum_read_malloc_memory,
+          gum_emit_malloc_ranges);
+
+      if (!info.carry_on)
+        return;
+    }
+  }
+}
+
+static void
+gum_emit_malloc_ranges (task_t task,
+                        void * user_data,
+                        unsigned type,
+                        vm_range_t * ranges,
+                        unsigned count)
+{
+  GumEnumerateMallocRangesContext * info =
+      (GumEnumerateMallocRangesContext *) user_data;
+  GumMemoryRange gum_range;
+  GumMallocRangeDetails details;
+  unsigned i;
+
+  if (!info->carry_on)
+    return;
+
+  details.range = &gum_range;
+
+  for (i = 0; i != count; i++)
+  {
+    vm_range_t range;
+    gboolean ret;
+
+    range = ranges[i];
+
+    gum_range.base_address = range.address;
+    gum_range.size = range.size;
+
+    ret = info->func (&details, info->user_data);
```

Comment:
Would be clearer to drop the `ret` variable here and just assign directly to `info->carry_on`.